parquet-hadoop/src/main/java/org/apache/parquet/hadoop/codec/NonBlockedDecompressor.java

```diff
@@ -0,0 +1,174 @@
+    return numBytes;
+  }
+
+  /**
+   * Sets input data for decompression.
+   * This should be called if and only if {@link #needsInput()} returns
+   * <code>true</code> indicating that more input data is required.
+   * (Both native and non-native versions of various Decompressors require
+   * that the data passed in via <code>b[]</code> remain unmodified until
+   * the caller is explicitly notified--via {@link #needsInput()}--that the
+   * buffer may be safely modified.  With this requirement, an extra
+   * buffer-copy can be avoided.)
+   *
+   * @param buffer Input data
+   * @param off    Start offset
+   * @param len    Length
+   */
+  @Override
+  public synchronized void setInput(byte[] buffer, int off, int len) {
+    SnappyUtil.validateBuffer(buffer, off, len);
+
+    if (inputBuffer.capacity() - inputBuffer.position() < len) {
+      final ByteBuffer newBuffer = ByteBuffer.allocateDirect(inputBuffer.position() + len);
+      inputBuffer.rewind();
+      newBuffer.put(inputBuffer);
+      final ByteBuffer oldBuffer = inputBuffer;
+      inputBuffer = newBuffer;
+      CleanUtil.cleanDirectBuffer(oldBuffer);
+    } else {
+      inputBuffer.limit(inputBuffer.position() + len);
+    }
+    inputBuffer.put(buffer, off, len);
+  }
+
+  @Override
+  public void end() {
+    CleanUtil.cleanDirectBuffer(inputBuffer);
+    CleanUtil.cleanDirectBuffer(outputBuffer);
+  }
+
+  @Override
+  public synchronized boolean finished() {
+    return finished && !outputBuffer.hasRemaining();
+  }
+
+  @Override
+  public int getRemaining() {
+    return 0;
+  }
+
+  @Override
+  public synchronized boolean needsInput() {
+    return !inputBuffer.hasRemaining() && !outputBuffer.hasRemaining();
+  }
+
+  @Override
+  public synchronized void reset() {
+    finished = false;
+    inputBuffer.rewind();
+    outputBuffer.rewind();
+    inputBuffer.limit(0);
+    outputBuffer.limit(0);
+  }
+
+  @Override
+  public boolean needsDictionary() {
+    return false;
+  }
+
+  @Override
+  public void setDictionary(byte[] b, int off, int len) {
+    // No-op		
+  }
+
+  /**
+   * Uncompress the content in the input buffer. The result is dumped to the
+   * specified output buffer.
+   *
+   * @param compressed   buffer[pos() ... limit()) containing the input data
+   * @param uncompressed output of the the uncompressed data. It uses buffer[pos()..]
+   * @return uncompressed data size
+   */
+  abstract protected int uncompress(ByteBuffer compressed, ByteBuffer uncompressed) throws IOException;
+
+  /**
+   * Get the uncompressed byte size of the given compressed input. This operation takes O(1) time.
+   *
+   * @param compressed input data [pos() ... limit())
+   * @param maxUncompressedLength maximum length of the uncompressed data
+   * @return uncompressed byte length of the given input
+   */
+  abstract protected int uncompressedLength(ByteBuffer compressed, int maxUncompressedLength) throws IOException;
+
+} //class NonBlockedDecompressor
```

Comment:
We generally don't need this ending comment

Reply:
Fixed